ublox_gps/src/ublox_firmware9.cpp

```diff
@@ -16,4 +17,113 @@ UbloxFirmware9::UbloxFirmware9(const std::string & frame_id, std::shared_ptr<dia
+    }
+  }
+
+  gps->setSaveOnShutdown(save_on_shutdown_);
+
+  //
+  // Configure the GNSS, only if the configuration is different
+  //
+  // First, get the current GNSS configuration
+  ublox_msgs::msg::CfgGNSS cfg_gnss;
+  if (gps->poll(cfg_gnss))
+  {
+    RCLCPP_DEBUG(node_->get_logger(), "Read GNSS config.");
+    RCLCPP_DEBUG(node_->get_logger(), "Num. tracking channels in hardware: %i", cfg_gnss.num_trk_ch_hw);
+    RCLCPP_DEBUG(node_->get_logger(), "Num. tracking channels to use: %i", cfg_gnss.num_trk_ch_use);
+  }
+  else
+  {
+    throw std::runtime_error("Failed to read the GNSS config.");
+  }
+
+  ublox_msgs::msg::CfgVALSET cfg_signal;
+  cfg_signal.layers = ublox_msgs::msg::CfgVALSET::LAYER_RAM;
+
+  using signal = ublox_msgs::msg::CfgVALSETCfgdata;
+  std::vector<signal> signalCfgData{};
+
+  // Configure GPS Signals
+  signalCfgData.push_back(generateSignalConfig_(signal::GPS_ENABLE, enable_gps_));
+  signalCfgData.push_back(generateSignalConfig_(signal::GPS_L1CA_ENABLE, enable_gps_));
+  if (gnss_->isSupported("GPS_L2C"))
+  {
+    signalCfgData.push_back(generateSignalConfig_(signal::GPS_L2C_ENABLE, enable_gps_));
+  }
+
+  // Configure SBAS Signals
+  signalCfgData.push_back(generateSignalConfig_(signal::SBAS_ENABLE, enable_gps_));
+  signalCfgData.push_back(generateSignalConfig_(signal::SBAS_L1CA_ENABLE, enable_gps_));
+
+  // Configure Galileo Signals
+  signalCfgData.push_back(generateSignalConfig_(signal::GAL_ENABLE, enable_galileo_));
+  signalCfgData.push_back(generateSignalConfig_(signal::GAL_E1_ENABLE, enable_galileo_));
+  if (gnss_->isSupported("GAL_E5B"))
+  {
+    signalCfgData.push_back(generateSignalConfig_(signal::GAL_E5B_ENABLE, enable_galileo_));
+  }
+
+  // Configure Beidou Signals
+  signalCfgData.push_back(generateSignalConfig_(signal::BDS_ENABLE, enable_beidou_));
+  signalCfgData.push_back(generateSignalConfig_(signal::BDS_B1_ENABLE, enable_beidou_));
+  if (gnss_->isSupported("BDS_B2"))
+  {
+    signalCfgData.push_back(generateSignalConfig_(signal::BDS_B2_ENABLE, enable_beidou_));
+  }
+
+  // Configure QZSS Signals
+  signalCfgData.push_back(generateSignalConfig_(signal::QZSS_ENABLE, enable_qzss_ && enable_gps_));
+  signalCfgData.push_back(generateSignalConfig_(signal::QZSS_L1CA_ENABLE, enable_qzss_ && enable_gps_));
+  if (gnss_->isSupported("QZSS_L2C"))
+  {
+    signalCfgData.push_back(generateSignalConfig_(signal::QZSS_L2C_ENABLE, enable_qzss_ && enable_gps_));
+  }
+
+  // Configure GLONASS Signals
+  signalCfgData.push_back(generateSignalConfig_(signal::GLO_ENABLE, enable_glonass_));
+  signalCfgData.push_back(generateSignalConfig_(signal::GLO_L1_ENABLE, enable_glonass_));
+  if (gnss_->isSupported("GLO_L2"))
+  {
+    signalCfgData.push_back(generateSignalConfig_(signal::GLO_L2_ENABLE, enable_glonass_));
+  }
+
+  cfg_signal.cfgdata.resize(signalCfgData.size());
+  for (size_t i = 0; i < signalCfgData.size(); ++i)
+  {
+    cfg_signal.cfgdata[i] = signalCfgData[i];
+  }
```

Comment:
Do we need to do this copy?  That is, can we just `push_back` onto `cfg_signal.cfgdata` vector as we need to above, and remove `signalCfgData` completely?

Reply:
Ah this is why we do code reviews.  I was oblivious that `cfg_signal.cfgdata` was a `std::vector`.  Should be fixed now.  This clean things up a bit.